Updated derivative rules for complex svd and pinverse

IvanYashchuk opened the pull request:
Updated `svd_backward` to work correctly for complex-valued inputs.
Updated `common_methods_invocations.py` to take dtype, device arguments for input construction.
Removed `test_pinverse` from `test_autograd.py`, it is replaced by entries to `common_methods_invocations.py`.
Added `svd` and `pinverse` to list of complex tests.

References for complex-valued SVD differentiation:

- https://giggleliu.github.io/2019/04/02/einsumbp.html
- https://arxiv.org/abs/1909.02659

The derived rules assume gauge invariance of loss functions, so the result would not be correct for loss functions that are not gauge invariant.
https://re-ra.xyz/Gauge-Problem-in-Automatic-Differentiation/

The same rule is implemented in Tensorflow and [BackwardsLinalg.jl](https://github.com/GiggleLiu/BackwardsLinalg.jl).

Ref. #33152

IvanYashchuk committed:
wip svd autograd update

IvanYashchuk committed:
Merge remote-tracking branch 'upstream/master' into autograd-svd

IvanYashchuk committed:
Added missing imaginary term when input is complex

IvanYashchuk committed:
Added tests; gradgradcheck doesn't work yet

IvanYashchuk committed:
Merge remote-tracking branch 'upstream/master' into autograd-svd

IvanYashchuk committed:
Updated tests

IvanYashchuk committed:
Added alternative imaginary term

IvanYashchuk committed:
Merge remote-tracking branch 'upstream/master' into autograd-svd

IvanYashchuk committed:
Fixed backward svd, finally all tests pass!

IvanYashchuk committed:
Remove unnecessary imag_term when only gv is defined

IvanYashchuk committed:
Fixed batched input case; chain_matmul works only for non-batched input

IvanYashchuk committed:
Enable complex common_methods tests

IvanYashchuk committed:
Added svd test cases that check grad for each returned tensor separately

IvanYashchuk committed:
Removed tmp tests from test_autograd.py; they're part of common_methods now

IvanYashchuk committed:
Enabled complex tests for pinverse

IvanYashchuk committed:
flake8

IvanYashchuk committed:
Added a note on gauge invariance requirement to docs

IvanYashchuk committed:
Remove commented line of code

IvanYashchuk committed:
Exclude type checks for svd grad tests

IvanYashchuk committed:
Merge remote-tracking branch 'upstream/master' into autograd-svd

IvanYashchuk committed:
Merge branch 'master' into autograd-svd

IvanYashchuk committed:
Merge branch 'master' into autograd-svd

IvanYashchuk commented on test/test_autograd.py:
```diff
@@ -4987,12 +4963,15 @@ def run_functional_checks(test_case, test_name, name, apply_fn, run_grad_checks,
+                'svd', 'pinverse'] + separate_complex_tests
 
 # this list corresponds to cases that are not currently implemented
 skip_cuda_list = ['bmm_complex', 'matmul_4d_4d_complex', 'inverse_batched_complex',
                   'solve_batched_broadcast_A_complex', 'solve_batched_broadcast_b_complex',
-                  'solve_batched_complex', 'solve_batched_dims_complex']
+                  'solve_batched_complex', 'solve_batched_dims_complex',
+                  'svd_batched_complex', 'svd_tall_all_batched_complex', 'svd_tall_batched_complex',
+                  'svd_wide_all_batched_complex', 'svd_wide_batched_complex', 'pinverse_batched_complex']
```
Remove this since batched matmul works for complex on CUDA now.

IvanYashchuk committed:
Merge remote-tracking branch 'upstream/master' into autograd-svd

IvanYashchuk committed:
Don't skip batched tests; complex matmul on cuda now works

IvanYashchuk committed:
Merge branch 'master' into autograd-svd

mruberry commented:
Hey @IvanYashchuk!

I'll let @albanD review the autograd aspects of this PR.

As for testing, I think this is a good opportunity to implement an OpInfo instead of using `method_tests()`. Yes, currently OpInfo's don't test that the function works with the jit properly, but @Lilyjjo's https://github.com/pytorch/pytorch/pull/47696 will soon change that.

Adding the OpInfo for these should be straightforward. You create a new instance of 

https://github.com/pytorch/pytorch/blob/5dfced3b0d716c885189914b7cdc4514621c4599/torch/testing/_internal/common_methods_invocations.py#L56

And add it to

https://github.com/pytorch/pytorch/blob/5dfced3b0d716c885189914b7cdc4514621c4599/torch/testing/_internal/common_methods_invocations.py#L210

@kshitij12345 has been doing a lot of work porting tests to use OpInfos and so may be helpful to consult.

The OpInfo's for svd and pinverse will need to implement this function:

https://github.com/pytorch/pytorch/blob/5dfced3b0d716c885189914b7cdc4514621c4599/torch/testing/_internal/common_methods_invocations.py#L117

Current OpInfos implement it indirectly by inheriting from a derived class of OpInfo, UnaryUfuncInfo. See here:

https://github.com/pytorch/pytorch/blob/5dfced3b0d716c885189914b7cdc4514621c4599/torch/testing/_internal/common_methods_invocations.py#L198

Once the OpInfo is implemented then the tests here:

https://github.com/pytorch/pytorch/blob/5dfced3b0d716c885189914b7cdc4514621c4599/test/test_ops.py#L51

Will be run automatically. 

There may be some limitations with the autograd tests that use OpInfos we need to work through. For example, I'm not sure if they support multiple inputs that require grad properly, but I think we should use this PR to improve those tests (since they're the future) vs. method_tests (since they're soon to be removed). 

Let me know if you have any questions. I can also help with changes to the autograd OpInfo tests.

IvanYashchuk commented:
Sure, that's right we shouldn't postpone moving to the new test system, somebody will have to do it anyway. I will add OpInfo tests, thanks for the instructions! I'll let you know once it's done.

albanD commented on torch/csrc/autograd/FunctionsManual.cpp:
```diff
@@ -1836,28 +1836,32 @@ Tensor svd_backward(const std::vector<torch::autograd::Variable> &grads, const T
   auto u = raw_u;
-  auto v = raw_v;
+  // TODO: remove .conj()
+  // once https://github.com/pytorch/pytorch/issues/45821 is resolved
```
From reading the issue, it is not clear if this is ever going to be fixed.
Maybe mention here that `torch.svd` returns the conjugate of what this function expect, hence the extra conj.

nit: you can move all the conj next to the todo to make it clearer which on  you're talking about in this comment. Or add a detailed comment at one place and reference to it from all the relevant places below.

IvanYashchuk replied:
I improved the comment by mentioning that currently SVD returns the conjugate of V.

I think I could resolve the issue soon, the changes should be minimal. It hasn't been fixed yet, because @antocuni  wanted to include it in his `torch.linalg.svd` PR, and then I guess it was decided not to do it there.

albanD replied:
But changing torch.svd would be a bc-breaking change right? So it would need to go through deprecation, etc?

IvanYashchuk replied:
The change concerns only the complex-valued inputs, the support for which was just recently added, and complex support is still in active development so it's not really bc-breaking. This issue was overlooked when complex support was added.

albanD replied:
Ok that should be fine then. I guess you can move all of these at the top so that they will be easy to remove when we update that. Thanks for the clarification!

albanD commented on torch/csrc/autograd/FunctionsManual.cpp:
```diff
@@ -1836,28 +1836,32 @@ Tensor svd_backward(const std::vector<torch::autograd::Variable> &grads, const T
     }
   }
-  auto vt = v.transpose(-2, -1);
+  auto vh = v.conj().transpose(-2, -1);
```
nit: you conjugate v above to conjugate it again here.
While I see that it matches the formulas, it looks a bit weird in this context.

Another option might be to name the input argument `conj_v` and work with that directly?

IvanYashchuk replied:
I agree that it looks weird, I wanted to keep those extra `.conj()` changes only at the beginning of the function with the hope it'll be fixed soon and keep everything else matching the formulas and so that it is less confusing.

albanD commented on torch/_torch_docs.py:
```diff
@@ -7833,6 +7835,9 @@ def merge_dicts(*dicts):
           from the forward pass is required for the backward operation.
 
+.. note:: With the complex-valued input the backward operation works correctly only
+          for gauge invariant loss functions.
```
nit: could you add a reference link that explains what "gauge invatiant loss functions" is here?

IvanYashchuk replied:
I added a link to https://re-ra.xyz/Gauge-Problem-in-Automatic-Differentiation/

albanD reviewed (commented):
Thanks for the PR!

I did not dive into the formula in details but the code make sense and gradcheck is happy so that is a good sign for me.

IvanYashchuk committed:
Merge remote-tracking branch 'upstream/master' into autograd-svd

IvanYashchuk committed:
Updated comment on torch.svd returning conj(V)

IvanYashchuk committed:
Added a link for gauge invariance topic

IvanYashchuk committed:
Added OpInfo-based tests

IvanYashchuk committed:
flake8

IvanYashchuk commented:
@mruberry I added OpInfo tests for svd and pinverse. It required slight changes to match the functionality of `method_tests`, particularly the possibility to provide a function that is applied to the output of `op` for gradcheck.
I didn't remove `method_tests` changes. Let's see first whether all CI tests pass.

IvanYashchuk committed:
Fixed typo in test_ops.py

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -44,12 +44,14 @@ def __init__(self, cls_name=None, test_name=None, *,
+    __slots__ = ['input', 'args', 'kwargs', 'output_process_fn_grad']
 
-    def __init__(self, input, *, args=tuple(), kwargs=None):
+    def __init__(self, input, *, args=tuple(), kwargs=None, output_process_fn_grad=None):
```
https://github.com/pytorch/pytorch/pull/48627 implemented similar functionality that I think this will be able to reuse. It should land soon.

IvanYashchuk replied:
I don't see how to reuse it. I need functionality to be able for the specific inputs define specific function to test by gradcheck. So that not `torch.svd(input)` is tested but `output_process_fn_grad(torch.svd(input))`.

IvanYashchuk replied:
Okay, I see `op.output_func` that does the same, but to use it here I'd need separate tests in `op_db` list. I think it's not possible to define different OpInfos for the same function because everything is tied to the `op.name` that has to be unique now.

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -219,7 +224,82 @@ def sample_inputs(self, device, dtype, requires_grad=False):
+    """Operator information for 'torch.svd'
+    """
+
+    def sample_inputs(self, device, dtype, requires_grad=False):
```
https://github.com/pytorch/pytorch/pull/48627 will provide a way to provide a custom sample_inputs that doesn't require defining a new class, which should simplify this code.

albanD reviewed (commented):
Looks good on my end (I haven't looked at the tests).

Let me know if I need to do another pass including the tests later.

IvanYashchuk committed:
Merge remote-tracking branch 'upstream/master' into autograd-svd

IvanYashchuk committed:
Added descriptions to sample_inputs functions for svd and pinverse

IvanYashchuk committed:
Added a link to github issue about test_unsupported_dtypes

IvanYashchuk committed:
Set supports_tensor_out=False

IvanYashchuk committed:
Fixed error due to merge

IvanYashchuk commented:
@mruberry I modified opinfo-based tests a bit. Now a separate class is not created, only the `sample_inputs` function.
All tests seem to pass.

IvanYashchuk committed:
Merge branch 'master' into autograd-svd

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -1695,30 +1797,65 @@ def method_tests():
-        ('svd', lambda dtype, device: random_fullrank_matrix_distinct_singular_value(S, 3, 3)[..., :, :(S - 2)], (False,),
-         'tall_all_batched', (), NO_ARGS, [skipCPUIfNoLapack, skipCUDAIfNoMagma],
-         lambda usv: (usv[0][..., :, :(S - 2)], usv[1], usv[2])),
+        # loss functions for complex-valued svd have to be "gauge invariant",
```
The OpInfo extension is really clever and looks great. Can we now remove these method test entries? That is, are these redundant with the OpInfo-based tests?

mruberry reviewed (commented):
Hey @IvanYashchuk!

The OpInfo extension looks great. My biggest question is if we can remove the method_test entries for these functions as redundant with the OpInfo-based tests?

IvanYashchuk committed:
Merge remote-tracking branch 'upstream/master' into autograd-svd

IvanYashchuk committed:
Removed svd and pinverse entries from methods_tests;

they are replaced with OpInfo-based tests with the same inputs and loss
functions

IvanYashchuk commented:
Sure, now when we see that all tests pass with the new OpInfo, we can remove old tests. I've done that.

mruberry reviewed (approved):
Cool! Nice work, @IvanYashchuk!

mruberry commented:
This LGTM to me but test failures are real. I think this may be enabling some behavior and the corresponding tests need to be updated to reflect that.

IvanYashchuk committed:
Merge remote-tracking branch 'upstream/master' into autograd-svd

IvanYashchuk committed:
Removed assertRaises for not implemented svd with complex dtype

IvanYashchuk commented:
Yes, this PR enabled autograd for norm with complex-valued input. I removed the RuntimeError assert, but I didn't update the tests further since it would require a bit more work than simply allowing the previously skipped `['nuc', 2, -2]` norm types. The current `test_norm_complex_autograd` tests whether the result of norm(complex_matrix) == norm(abs(complex_matrix)), but it can't be true in general (I think so, and computations prove that, couldn't find proper references). I updated the TODO comment.

mruberry commented:
> Yes, this PR enabled autograd for norm with complex-valued input. I removed the RuntimeError assert, but I didn't update the tests further since it would require a bit more work than simply allowing the previously skipped `['nuc', 2, -2]` norm types. The current `test_norm_complex_autograd` tests whether the result of norm(complex_matrix) == norm(abs(complex_matrix)), but it can't be true in general (I think so, and computations prove that, couldn't find proper references). I updated the TODO comment.

That seems reasonable for this PR if @kurtamohler's onboard with the follow-up. @kurtamohler?

kurtamohler commented:
That sounds good to me!

mruberry commented on torch/_torch_docs.py:
```diff
@@ -8000,7 +8000,7 @@ def merge_dicts(*dicts):
 
 This function returns a namedtuple ``(U, S, V)`` which is the singular value
-decomposition of a input real matrix or batches of real matrices :attr:`input` such that
+decomposition of a input matrix or batches of matrices :attr:`input` such that
```
Future work: "of an input matrix or batch of matrices"

mruberry commented on torch/_torch_docs.py:
```diff
@@ -8012,6 +8012,8 @@ def merge_dicts(*dicts):
 If :attr:`compute_uv` is ``False``, the returned `U` and `V` matrices will be zero matrices
 of shape :math:`(m \times m)` and :math:`(n \times n)` respectively. :attr:`some` will be ignored here.
 
+Supports real-valued and complex-valued input.
```
Future work: specify the dtypes explicitly since "real-valued" would be every non-bool non-complex dtype

mruberry commented on torch/_torch_docs.py:
```diff
@@ -8036,6 +8038,9 @@ def merge_dicts(*dicts):
 .. note:: When :attr:`compute_uv` = ``False``, backward cannot be performed since `U` and `V`
           from the forward pass is required for the backward operation.
 
+.. note:: With the complex-valued input the backward operation works correctly only
```
Future work: this is too vague and referencing other documentation is never ideal

mruberry commented:
Let's snap this PR here to keep things moving. Going forward we should look at doc refinements for torch.linalg.svd and torch.linalg.pinv and update the torch.linalg.norm testing, as discussed above.

mruberry commented:
Hey @IvanYashchuk, bad news / interesting news. I had to unland this because it caused a significant increase in test timings, like adding 30 minutes to some test builds. That's the bad news.

The interesting news is now we get to have a discussion about testing thoroughness vs. cost.

cc @malfet, @ngimel, @albanD, @kurtamohler 

For example, here's a CI build after this PR:

https://app.circleci.com/pipelines/github/pytorch/pytorch/252033/workflows/b9605b69-1006-4c0e-a49f-e89f85a469ad/jobs/9675634

And here's the same build before it: 

https://app.circleci.com/pipelines/github/pytorch/pytorch/251979/workflows/dcf1c55f-a041-497d-a1c0-11c0e8d153f6/jobs/9674037

Is this discrepancy really caused by this PR? Perhaps surprisingly: yes.

While most of the tests this PR adds run quickly, there are some standouts:

```
test_method_grad_svd_cuda_complex128 (__main__.TestGradientsCUDA) ... ok (95.033s)
test_method_grad_svd_cuda_float64 (__main__.TestGradientsCUDA) ... ok (47.357s)
test_method_gradgrad_svd_cuda_complex128 (__main__.TestGradientsCUDA) ... ok (294.281s)
test_method_gradgrad_svd_cuda_float64 (__main__.TestGradientsCUDA) ... ok (99.565s)
test_fn_gradgrad_svd_cuda_complex128 (__main__.TestGradientsCUDA) ... ok (307.492s)
test_fn_gradgrad_svd_cuda_float64 (__main__.TestGradientsCUDA) ... ok (104.823s)
test_fn_grad_svd_cuda_complex128 (__main__.TestGradientsCUDA) ... ok (93.672s)
test_fn_grad_svd_cuda_float64 (__main__.TestGradientsCUDA) ... ok (47.417s)
```

For just CUDA grad and gradgrad testing of SVD that's almost 20 minutes! Other tests and the pinverse tests are much faster.

Here's some of the tests from above running on CPU, for example:
```
test_method_gradgrad_svd_cpu_complex128 (__main__.TestGradientsCPU) ... ok (37.549s)
test_method_gradgrad_svd_cpu_float64 (__main__.TestGradientsCPU) ... ok (10.695s)
```

While these tests aren't super fast, they take less than a minute compared to the corresponding CUDA tests that over six minutes.

A few things here:

- it's clear we have to be smarter about what we're testing vs. the cost of testing it
- a check for build time regressions in the Github CI would be nice so people don't have to compare before and after timings for different builds regularly, since that's time-consuming

Some options for being smarter:

- We could stop testing method grad and gradgrad and rely on function grad and gradgrad tests (@albanD) 
- We can stop testing svd grad and gradgrad on CUDA, if `svd_backward` is a composite operation (and a cursory look suggests it is) without custom CUDA bits then this seems reasonable given how long it takes
- We could limit SVD's sample inputs

I'd like to suggest starting with the first two options. Maybe we should stop testing the CUDA grad and gradgrad of other functions whose backward is a composite operation, too?

Separately, we should follow-up on CUDA SVD performance. Luckily @xwang233 already has a PR that may address some of these issues: https://github.com/pytorch/pytorch/pull/48436.

rgommers commented:
> a check for build time regressions in the Github CI would be nice so people don't have to compare before and after timings for different builds regularly, since that's time-consuming

That seems hard to do reliably. Normally you'd notice this anyway when adding the tests, since you're running only those tests when developing. So perhaps best to do as a combination of setting expectations and once in a while cleaning up outliers? Easy to find outliers:
```
$ pytest test/test_nn.py --durations 10
...
17.34s call     test/test_nn.py::TestNN::test_conv_double_backward
10.99s call     test/test_nn.py::TestNNInit::test_trunc_normal
8.08s call     test/test_nn.py::TestNN::test_TransformerEncoderLayer_relu_activation
6.33s call     test/test_nn.py::TestNNDeviceTypeCPU::test_triplet_margin_with_distance_loss_cpu
5.77s call     test/test_nn.py::TestNN::test_LocalResponseNorm_3d_custom_params
5.34s call     test/test_nn.py::TestNN::test_FractionalMaxPool3d_asymsize
5.33s call     test/test_nn.py::TestNN::test_grid_sample
5.22s call     test/test_nn.py::TestNNDeviceTypeCPU::test_grid_sample_large_index_2d_cpu_float64
4.63s call     test/test_nn.py::TestNN::test_interpolate_nearest_scale_3d
4.12s call     test/test_nn.py::TestNN::test_l1_loss_correct
```

mruberry commented:
> > a check for build time regressions in the Github CI would be nice so people don't have to compare before and after timings for different builds regularly, since that's time-consuming
> 
> That seems hard to do reliably. Normally you'd notice this anyway when adding the tests, since you're running only those tests when developing. 

I don't know if it is so challenging. What if we diffed the test files produced for new tests, for example, and summed their total time and reported if it was over a minute?

rgommers commented:
> I don't know if it is so challenging. What if we diffed the test files produced for new tests, for example, and summed their total time and reported if it was over a minute?

You will have the same problem that codecov has, which is that diffing against a baseline run under different circumstances or which may be further from the branch point of your PR can be wrong or nonexisting for lots of reasons. I'd expect the false positive rate to be high.

malfet commented:
@mruberry thank you very much for the revert. Please note, that adding all those tests pushed pytorch_linux_xenial_cuda11_1_cudnn8_py3_gcc7_test total test time beyond 5h limit. We can always shard it to test1 and test2, but it would be good to understand why those tests need to be so slow.

albanD commented:
I do agree with Mike that for such cases, it is ok to skip the grad and gradgrad check on cuda.
Or maybe do only one with a 2x2 matrix to make sure it actually runs.

> We could stop testing method grad and gradgrad and rely on function grad and gradgrad tests

You mean that right now, when an op is both a method and a function, we (grad)gradcheck both?

mruberry commented:
> I do agree with Mike that for such cases, it is ok to skip the grad and gradgrad check on cuda.
> Or maybe do only one with a 2x2 matrix to make sure it actually runs.
> 
> > We could stop testing method grad and gradgrad and rely on function grad and gradgrad tests
> 
> You mean that right now, when an op is both a method and a function, we (grad)gradcheck both?

Yes. We currently re-run the grad and grad grad checks for the function variant and the method variant of each operation. Seems like we can stop running them on the method variant?

albanD commented:
For all the ops that are automatically bound to python via codegen from `native_functions.yaml`, we can definitely do that yes.

mruberry commented:
@ngimel and I looked at the current test timings for SVD CUDA backward, and the existing tests take about 5 minutes. By creating an OpInfo there's a 4x expansion of these tests ({function, method} x {double, cdouble}). Removing the method variants would still create a 2x expansion. @ngimel mentioned she'd like the opportunity to review SVD's CUDA backwards before we decide on a course of action.

mruberry commented:
Thanks to @ngimel and @albanD we're going to stop testing method grad and gradgrad for the moment and suggest we stop testing CUDA svd gradgrad. @IvanYashchuk, sorry to ask you to make another revision of this PR, but would you add skips for svd's CUDA gradgrad's tasks? 

With these changes we expect svd's CUDA gradient test time to be lower than in builds today.

IvanYashchuk committed:
Merge remote-tracking branch 'upstream/master' into autograd-svd

IvanYashchuk committed:
Skip cuda gradchecks

IvanYashchuk committed:
Mark test_fn_gradgrad as slowtest

IvanYashchuk committed:
Use smaller arrays for tests

IvanYashchuk commented:
@mruberry I added the skips for cuda gradchecks. I also marked gradgrad test as slow.
I removed one test case and changed batch size from `3` and `3x3` to `2` 18651b4.
Here are the timings for my local tests before changing the dimensions
```
217.71s call     test/test_ops.py::TestGradientsCUDA::test_fn_grad_svd_cuda_complex128
101.21s call     test/test_ops.py::TestGradientsCUDA::test_fn_grad_svd_cuda_float64
9.02s call     test/test_ops.py::TestGradientsCPU::test_fn_grad_svd_cpu_complex128
4.67s call     test/test_ops.py::TestGradientsCPU::test_fn_grad_svd_cpu_float64
747.57s call     test/test_ops.py::TestGradientsCUDA::test_fn_gradgrad_svd_cuda_complex128
251.72s call     test/test_ops.py::TestGradientsCUDA::test_fn_gradgrad_svd_cuda_float64
150.30s call     test/test_ops.py::TestGradientsCPU::test_fn_gradgrad_svd_cpu_complex128
43.29s call     test/test_ops.py::TestGradientsCPU::test_fn_gradgrad_svd_cpu_float64
```
and after
```
26.10s call     test/test_ops.py::TestGradientsCUDA::test_fn_grad_svd_cuda_complex128
12.88s call     test/test_ops.py::TestGradientsCUDA::test_fn_grad_svd_cuda_float64
3.34s call     test/test_ops.py::TestGradientsCPU::test_fn_grad_svd_cpu_complex128
1.76s call     test/test_ops.py::TestGradientsCPU::test_fn_grad_svd_cpu_float64
178.80s call     test/test_ops.py::TestGradientsCUDA::test_fn_gradgrad_svd_cuda_complex128
58.08s call     test/test_ops.py::TestGradientsCPU::test_fn_gradgrad_svd_cpu_complex128
56.15s call     test/test_ops.py::TestGradientsCUDA::test_fn_gradgrad_svd_cuda_float64
16.74s call     test/test_ops.py::TestGradientsCPU::test_fn_gradgrad_svd_cpu_float64
```

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -800,6 +886,27 @@ def sample_inputs(self, device, dtype, requires_grad=False):
+               # cuda gradchecks are very slow
+               # see discussion https://github.com/pytorch/pytorch/pull/47761#issuecomment-747316775
+               SkipInfo('TestGradients', 'test_fn_gradgrad', device_type='cuda'),
+               SkipInfo('TestGradients', 'test_fn_grad', device_type='cuda'))),
```
I think we can keep the grad checks going, let's just exclude the gradgrad checks on CUDA for now

IvanYashchuk replied:
Should CUDA gradgrad check be always skipped or tested only with `PYTORCH_TEST_WITH_SLOW`?

mruberry replied:
Let's just unconditionally skip them on svd for now. @albanD and I are going to look at our test matrix more closely in January.

mruberry commented on test/test_jit.py:
```diff
@@ -15522,7 +15522,11 @@ def forward(self, x):
     'test_slogdet_batched_symmetric',
     'test_slogdet_batched_symmetric_pd',
-    'test_slogdet_batched_distinct_singular_values'
+    'test_slogdet_batched_distinct_singular_values',
```
Are these necessary anymore?

IvanYashchuk replied:
No, I'll remove it.

mruberry commented:
New test timings look good. Let's keep CUDA grad for now and just skip CUDA gradgrad. Then let's merge this.

IvanYashchuk committed:
Undo changes to test/test_jit.py

IvanYashchuk committed:
Don't skip cuda grad checks

IvanYashchuk commented:
Done. Let's try merging it.
